server.js

```diff
@@ -785,41 +797,97 @@ cache(function (data, match, sendBadge, request) {
+ *  - examples: Example URIs for this service. These should use the format
+ *    specified in `uri`, and can be used to demonstrate how to use badges for
+ *    this service.
+ *  - handler: Asynchronous function to handle requests for this service. This
+ *    function takes two arguments:
+ *    1. Capture groups from the `uri` regular expression
+ *    2. A `request` method for sending cached HTTP requests to third-party
+ *       servers.
+ */
+function registerService(config) {
+  // Regular expressions treat "/" specially, so we need to escape them
+  const escapedPath = config.uri.replace(/\//g, '\\/');
+  const fullRegex = '^' + escapedPath + '.(svg|png|gif|jpg|json)$';
+
+  camp.route(new RegExp(fullRegex),
+  cache(function(data, match, sendBadge, request) {
+    // Assumes the final capture group is the extension
+    const format = match.pop();
+    // Discard the first capture group as it's the entire match
+    const params = match.slice(1);
+
+    const badgeData = getBadgeData(config.type, data);
+    config.handler(params, request.asPromise)
+      .then(serviceData => {
+        let text = badgeData.text;
+        if (serviceData.text) {
+          text[1] = serviceData.text;
+        }
+        Object.assign(badgeData, serviceData);
+        badgeData.text = text;
+        sendBadge(format, badgeData);
+      })
+      .catch(error => {
+        console.error(error);
+        const text = badgeData.text;
+        text[1] = 'error';
+        badgeData.text = text;
+        sendBadge(format, badgeData);
+      });
+  }));
+}
+
 // AppVeyor CI integration.
-camp.route(/^\/appveyor\/ci\/([^\/]+\/[^\/]+)(?:\/(.+))?\.(svg|png|gif|jpg|json)$/,
-cache(function(data, match, sendBadge, request) {
-  var repo = match[1];  // eg, `gruntjs/grunt`.
-  var branch = match[2];
-  var format = match[3];
-  var apiUrl = 'https://ci.appveyor.com/api/projects/' + repo;
-  if (branch != null) {
-    apiUrl += '/branch/' + branch;
-  }
-  var badgeData = getBadgeData('build', data);
-  request(apiUrl, { headers: { 'Accept': 'application/json' } }, function(err, res, buffer) {
-    if (err != null) {
-      badgeData.text[1] = 'inaccessible';
-      sendBadge(format, badgeData);
+registerService({
+  name: 'AppVeyor',
+  type: 'build',
+  uri: '/appveyor/ci/([^/]+/[^/]+)(?:/(.+))?',
+  examples: [
+    {uri: '/appveyor/ci/gruntjs/grunt'},
+    {
+      name: 'Branch',
+      params: '/appveyor/ci/gruntjs/grunt/master',
+    },
+  ],
+  handler: async ([repo, branch], request) => {
+    let apiUrl = 'https://ci.appveyor.com/api/projects/' + repo;
+    if (branch != null) {
+      apiUrl += '/branch/' + branch;
     }
+
+    let result;
     try {
-      var data = JSON.parse(buffer);
-      var status = data.build.status;
+      result = await request(apiUrl, {
+        headers: { 'Accept': 'application/json' }
+      });
+    } catch (e) {
+      return {text: 'inaccessible'};
+    }
```

Comment:
Actually I just remembered that I do already have a catch block in the `registerService`, so I don't need to have one here too 😛